node/impl/full/eth_trace.go

```diff
@@ -38,217 +37,441 @@ func decodePayload(payload []byte, codec uint64) (ethtypes.EthBytes, error) {
+	sender, err := lookupEthAddress(from, st)
 	if err != nil {
-		log.Warnf("buildTraces: failed to lookup from address %s: %v", et.Msg.From, err)
-		return nil
+		return nil, xerrors.Errorf("top-level message sender %s s could not be found: %w", from, err)
 	}
-	to, err := lookupEthAddress(et.Msg.To, st)
+	return &environment{caller: sender}, nil
+}
+
+func traceToAddress(act *types.ActorTrace) ethtypes.EthAddress {
+	if act.State.Address != nil {
+		if addr, err := ethtypes.EthAddressFromFilecoinAddress(*act.State.Address); err == nil {
+			return addr
+		}
+	}
+	return ethtypes.EthAddressFromActorID(act.Id)
+}
+
+// buildTraces recursively builds the traces for a given ExecutionTrace by walking the subcalls
+func buildTraces(env *environment, addr []int, et *types.ExecutionTrace) error {
+	trace, recurseInto, err := buildTrace(env, addr, et)
 	if err != nil {
-		log.Warnf("buildTraces: failed to lookup to address %s: %w", et.Msg.To, err)
+		return xerrors.Errorf("at trace %v: %w", addr, err)
+	}
+
+	if trace != nil {
+		env.traces = append(env.traces, trace)
+		env.subtraceCount++
+	}
+
+	// Skip if there's nothing more to do and/or `buildTrace` told us to skip this one.
+	if recurseInto == nil || recurseInto.InvokedActor == nil || len(recurseInto.Subcalls) == 0 {
 		return nil
 	}
 
-	// Skip the trace if we never reached the point where we invoked this actor.
+	subEnv := &environment{
+		caller: traceToAddress(recurseInto.InvokedActor),
+		isEVM:  builtinactors.IsEvmActor(recurseInto.InvokedActor.State.Code),
+		traces: env.traces,
+	}
+	addr = addr[:len(addr):len(addr)]
+	for i := range recurseInto.Subcalls {
+		err := buildTraces(subEnv, append(addr, subEnv.subtraceCount), &recurseInto.Subcalls[i])
+		if err != nil {
+			return err
+		}
+	}
+	trace.Subtraces = subEnv.subtraceCount
+	env.traces = subEnv.traces
+
+	return nil
+}
+
+func decodeCreate(msg *types.MessageTrace) (callType string, initcode []byte, err error) {
+	switch msg.Method {
+	case builtin.MethodsEAM.Create:
+		params, err := decodeParams[eam.CreateParams](msg)
+		if err != nil {
+			return "", nil, err
+		}
+		return "create", params.Initcode, nil
+	case builtin.MethodsEAM.Create2:
+		params, err := decodeParams[eam.Create2Params](msg)
+		if err != nil {
+			return "", nil, err
+		}
+		return "create2", params.Initcode, nil
+	case builtin.MethodsEAM.CreateExternal:
+		input, err := decodePayload(msg.Params, msg.ParamsCodec)
+		if err != nil {
+			return "", nil, err
+		}
+		return "create", input, nil
+	default:
+		return "", nil, xerrors.Errorf("unexpected CREATE method %d", msg.Method)
+	}
+}
+
+// buildTrace processes the passed execution trace and updates the environment, if necessary.
+//
+// On success, it returns a trace to add (or nil to skip) and the trace recurse into (or nil to skip).
+func buildTrace(env *environment, addr []int, et *types.ExecutionTrace) (*ethtypes.EthTrace, *types.ExecutionTrace, error) {
```

Comment:
this function is really long and complex and we should consider refactoring into handlers for different code paths